src/SourceFile.js

```diff
@@ -119,91 +148,82 @@ class SourceFile {
      * The content of the file, stored as regular Javascript string
      * encoded in UTF-8.
      *
-     * @type {String}
+     * @type {String|undefined}
      * @protected
      */
     content;
-    
+
     /**
      * Get the content of this file encoded as a regular Javascript
      * string.
-     * @returns {String} the content of the file, encoded as a JS string
+     * @returns {String|undefined} the content of the file, encoded as a JS string
      */
     getContent() {
         if (this.content === undefined) {
             this.read();
         }
         return this.content;
     }
-    
+
     /**
      * Get the content of this file, encoded as an array of lines. Each
      * line has its trailing newline character removed.
-     * @returns {Array.<String>} the content as an array of lines
+     * @returns {Array.<String>|undefined} the content as an array of lines
      */
     getLines() {
-        if (this.content === undefined) {
+        if (typeof(this.content) === 'undefined') {
             this.read();
         }
-        return this.content.split(/[\r\n]+/g);
-    }
-
-    /**
-     * Set the content of this file to the given array of lines. Each
-     * line should not have a trailing newline character. The file
-     * remains modified in-memory only and is not written out to disk
-     * until the {@link "SourceFile#write"} method is called. The newlines
-     * (LF only) are re-added before the contents are written back to disk.
-     * @param {Array.<String>} lines the lines to set as the content
-     */
-    setLines(lines) {
-        if (Array.isArray(lines)) {
-            this.content = lines.join('\n');
-            this.raw = Buffer.from(this.content, "utf8");
-            this.dirty = true;
-        }
+        return this.content?.split(/[\r\n]+/g);
     }
 
     /**
-     * The current length of the file, including any modifications.
+     * The current length of the file content, including any modifications.
+     * Note that this is the length of the content in Unicode characters,
+     * not the length of the raw bytes.
+     *
      * @returns {Number} the length in Unicode characters of this file
      */
     getLength() {
         if (this.content === undefined) {
             this.read();
         }
-        return this.content.length;
+        return this.content?.length || 0;
     }
 
     /**
-     * The type of this file.
+     * The type of this file. This should match the type of
+     * pipeline elements that process this same type of file.
+     *
      * @type {String}
      * @protected
      */
-    type;
-    
+    type = "";
```

Comment:
I think type should be required in the ctor, rather than providing an implicit default which is ultimately meaningless (unless we want to allow empty type as a special "universal" case?)

Reply:
The type is not required in the constructor, and so we had to have a default. I think the empty type should be "universal", yes.

Reply:
Could you please document the universal type?